src/server/rdb_load.cc

```diff
@@ -883,46 +862,66 @@ void RdbLoaderBase::OpaqueObjLoader::CreateZSet(const LoadTrace* ltrace) {
+  stream* s;
+  if (config_.append) {
+    if (!EnsureObjEncoding(OBJ_STREAM, OBJ_ENCODING_STREAM)) {
+      return;
+    }
 
-  stream* s = streamNew();
+    s = static_cast<stream*>(pv_->RObjPtr());
+  } else {
+    s = streamNew();
+  }
+
+  auto cleanup = absl::Cleanup([&] {
+    if (config_.append) {
+      freeStream(s);
+    }
+  });
 
-  auto cleanup = absl::Cleanup([&] { freeStream(s); });
+  for (size_t i = 0; i < ltrace->arr.size(); i += 2) {
+    string_view nodekey = ToSV(ltrace->arr[i].rdb_var);
+    string_view data = ToSV(ltrace->arr[i + 1].rdb_var);
 
-  for (const auto& seg : ltrace->arr) {
-    for (size_t i = 0; i < seg.size(); i += 2) {
-      string_view nodekey = ToSV(seg[i].rdb_var);
-      string_view data = ToSV(seg[i + 1].rdb_var);
+    uint8_t* lp = (uint8_t*)data.data();
 
-      uint8_t* lp = (uint8_t*)data.data();
+    if (!streamValidateListpackIntegrity(lp, data.size(), 0)) {
+      LOG(ERROR) << "Stream listpack integrity check failed.";
+      ec_ = RdbError(errc::rdb_file_corrupted);
+      return;
+    }
+    unsigned char* first = lpFirst(lp);
+    if (first == NULL) {
+      /* Serialized listpacks should never be empty, since on
+       * deletion we should remove the radix tree key if the
+       * resulting listpack is empty. */
+      LOG(ERROR) << "Empty listpack inside stream";
+      ec_ = RdbError(errc::rdb_file_corrupted);
+      return;
+    }
+    uint8_t* copy_lp = (uint8_t*)zmalloc(data.size());
+    ::memcpy(copy_lp, lp, data.size());
+    /* Insert the key in the radix tree. */
+    int retval =
+        raxTryInsert(s->rax_tree, (unsigned char*)nodekey.data(), nodekey.size(), copy_lp, NULL);
+    if (!retval) {
+      zfree(copy_lp);
+      LOG(ERROR) << "Listpack re-added with existing key";
+      ec_ = RdbError(errc::rdb_file_corrupted);
+      return;
+    }
+  }
 
-      if (!streamValidateListpackIntegrity(lp, data.size(), 0)) {
-        LOG(ERROR) << "Stream listpack integrity check failed.";
-        ec_ = RdbError(errc::rdb_file_corrupted);
-        return;
-      }
-      unsigned char* first = lpFirst(lp);
-      if (first == NULL) {
-        /* Serialized listpacks should never be empty, since on
-         * deletion we should remove the radix tree key if the
-         * resulting listpack is empty. */
-        LOG(ERROR) << "Empty listpack inside stream";
-        ec_ = RdbError(errc::rdb_file_corrupted);
-        return;
-      }
-      uint8_t* copy_lp = (uint8_t*)zmalloc(data.size());
-      ::memcpy(copy_lp, lp, data.size());
-      /* Insert the key in the radix tree. */
-      int retval =
-          raxTryInsert(s->rax_tree, (unsigned char*)nodekey.data(), nodekey.size(), copy_lp, NULL);
-      if (!retval) {
-        zfree(copy_lp);
-        LOG(ERROR) << "Listpack re-added with existing key";
-        ec_ = RdbError(errc::rdb_file_corrupted);
-        return;
-      }
+  // We only load the stream_trace on the final read, so if not read we
```

Comment:
I did not understand this comment. Can you explain please?

Reply:
i've updated the comment

what I mean is `ReadStreams` is split into two sections:
- Reading the stream entries (`ltrace->arr`)
- Reading the stream metadata and consumer groups (`ltrace->stream_trace`)

loading the stream metadata and consumer groups in partial reads would be quite complex, and i'm guessing isn't expected to be large enough to require partial reads? so wasn't sure if it's worth trying to load consumer groups in partial reads?

the simplest option seems to be just load the stream entries (`ltrace->arr`) in partial reads, then on the final read also read the stream metadata and consumer groups (`ltrace->stream_trace`)